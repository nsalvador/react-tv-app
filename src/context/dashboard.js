import React, { useContext } from 'react';

const DashboardContext = React.createContext();

export default DashboardContext;

export const useDashboardContext = () => useContext(DashboardContext);
