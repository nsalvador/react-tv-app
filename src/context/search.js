import React, { useContext } from 'react';

const SearchContext = React.createContext();

export default SearchContext;

export const useSearchContext = () => useContext(SearchContext);
