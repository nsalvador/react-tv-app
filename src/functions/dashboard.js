import { useState, useEffect, useReducer } from 'react';
import { database } from '../firebase';
import subscriptionsReducer from '../reducers/subscriptions';

const useDashboard = (id) => {
	const [subscriptions, dispatch] = useReducer(subscriptionsReducer, []);
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState(false);

	useEffect(() => {
		setLoading(true);
		database
			.ref(`users/${id}/shows`)
			.once(`value`)
			.then((dataSnapshot) => {
				const subscriptions = [];
				dataSnapshot.forEach((childSnapshot) => {
					subscriptions.push({
						id: childSnapshot.key,
						...childSnapshot.val(),
					});
				});
				setLoading(false);
				dispatch({ type: 'POPULATE_SUBSCRIPTIONS', subscriptions });
			})
			.catch((err) => setError(err));
	}, [id]);

	return { error, loading, subscriptions, dispatch };
};

export default useDashboard;
