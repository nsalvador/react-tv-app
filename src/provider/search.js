import React, { useState } from 'react';

import SearchContext from '../context/search';

const SearchProvider = (props) => {
	const [shows, SET_SHOWS] = useState([]);
	const [error, SET_ERROR] = useState('');
	const [isLoading, SET_LOADING] = useState(false);
	const value = { shows, SET_SHOWS, error, SET_ERROR, isLoading, SET_LOADING };

	return <SearchContext.Provider value={value} {...props} />;
};

export default SearchProvider;
