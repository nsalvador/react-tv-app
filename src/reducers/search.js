const searchReducer = (state, action) => {
	switch (action.type) {
		case 'SET_SHOWS':
			return [...action.shows];
		default:
			return state;
	}
};

export default searchReducer;
