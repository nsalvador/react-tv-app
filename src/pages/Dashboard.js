import React from 'react';
import { Link } from 'react-router-dom';

import Layout from '../components/Layout';
import { useDashboardContext } from '../context/dashboard';

const DashboardPage = () => {
	const { subscriptions } = useDashboardContext();

	return (
		<div>
			<Layout>
				{subscriptions.length !== 0 ? (
					<ol>
						{subscriptions.map((show) => (
							<li key={show.id}>
								{show.seriesName}
								<button>-</button>
							</li>
						))}
					</ol>
				) : (
					<p>
						No Subscriptions. Start a <Link to="/search">Search</Link>
					</p>
				)}
			</Layout>
		</div>
	);
};

export default DashboardPage;
