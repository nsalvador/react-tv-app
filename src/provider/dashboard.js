import React from 'react';

import { database } from '../firebase';
import useDashboard from '../functions/dashboard';
import DashboardContext from '../context/dashboard';
import { useAuthContext } from '../context/auth';

const DashboardProvider = (props) => {
	const { user } = useAuthContext();
	const { subscriptions, dispatch, loading: isLoading } = useDashboard(
		user?.uid
	);

	const remove = (show) => {
		const { uid } = user;
		const { id } = show;
		database
			.ref(`users/${uid}/shows/${id}`)
			.remove()
			.then(() => dispatch({ type: 'REMOVE_SHOW', id }));
	};

	const add = (show) => {
		const { uid } = user;
		const { id } = show;
		if (!subscriptions.some((item) => item.id === id)) {
			database
				.ref(`users/${uid}/shows`)
				.child(`${id}`)
				.set(show)
				.then(() => dispatch({ type: 'ADD_SHOW', show }));
		}
	};

	return (
		<DashboardContext.Provider
			value={{ subscriptions, dispatch, isLoading, remove, add }}
			{...props}
		/>
	);
};

export default DashboardProvider;
