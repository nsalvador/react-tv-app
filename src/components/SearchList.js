import React from 'react';

import { useDashboardContext } from '../context/dashboard';
import { useSearchContext } from '../context/search';

const SearchList = () => {
	const { error, shows } = useSearchContext();
	const { dispatch } = useDashboardContext();

	return (
		<div>
			{error ? (
				<p>{error}</p>
			) : (
				<ol>
					{shows.map((show) => (
						<li key={show.id}>
							{show.seriesName}
							<button onClick={() => dispatch({ type: 'ADD_SHOW', show })}>
								+
							</button>
						</li>
					))}
				</ol>
			)}
		</div>
	);
};

export default SearchList;
