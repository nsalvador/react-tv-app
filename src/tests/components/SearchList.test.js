import React from 'react';
import { shallow } from 'enzyme';

import * as DashboardContext from '../../context/dashboard';
import * as SearchContext from '../../context/search';
import SearchList from '../../components/SearchList';
import shows from '../fixtures/shows';

describe('Component: <SearchList />', () => {
	it('1. Should render component when search is unsuccessful', () => {
		const mockDispatch = jest.fn();
		const error = 'Not Found';
		const shows = [];
		jest
			.spyOn(SearchContext, 'useSearchContext')
			.mockImplementation(() => ({ error, shows }));
		jest
			.spyOn(DashboardContext, 'useDashboardContext')
			.mockImplementation(() => ({ dispatch: mockDispatch }));
		const wrapper = shallow(<SearchList />);
		expect(wrapper).toMatchSnapshot();
		const p = wrapper.find('p');
		expect(p.text()).toBe(error);
	});

	it('2. Should render component when search is successful', () => {
		const mockDispatch = jest.fn();
		const error = '';
		jest
			.spyOn(SearchContext, 'useSearchContext')
			.mockImplementation(() => ({ error, shows }));
		jest
			.spyOn(DashboardContext, 'useDashboardContext')
			.mockImplementation(() => ({ dispatch: mockDispatch }));

		const wrapper = shallow(<SearchList />);
		expect(wrapper).toMatchSnapshot();

		const ol = wrapper.find('ol');
		expect(ol.children()).toHaveLength(shows.length);

		const showNames = wrapper.find('li').map((node) => {
			const nodeShowName = node.text();
			return nodeShowName.substr(0, nodeShowName.length - 1);
		});
		expect(showNames).toEqual(shows.map(({ seriesName }) => seriesName));
	});
});
