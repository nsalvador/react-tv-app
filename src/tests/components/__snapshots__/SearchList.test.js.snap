// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <SearchList /> 1. Should render component when search is unsuccessful 1`] = `
<div>
  <p>
    Not Found
  </p>
</div>
`;

exports[`Component: <SearchList /> 2. Should render component when search is successful 1`] = `
<div>
  <ol>
    <li
      key="281662"
    >
      Marvel's Daredevil
      <button
        onClick={[Function]}
      >
        +
      </button>
    </li>
    <li
      key="249135"
    >
      Daredevils of the Red Circle
      <button
        onClick={[Function]}
      >
        +
      </button>
    </li>
    <li
      key="113851"
    >
      Daredevils
      <button
        onClick={[Function]}
      >
        +
      </button>
    </li>
    <li
      key="274196"
    >
      American Daredevils
      <button
        onClick={[Function]}
      >
        +
      </button>
    </li>
    <li
      key="244691"
    >
      Super Dave: Daredevil For Hire
      <button
        onClick={[Function]}
      >
        +
      </button>
    </li>
  </ol>
</div>
`;
